### PR TITLE
Increase field count limit to 3 for promoting a struct with no field accesses.

### PR DESCRIPTION
**Pull request by sivarv**
This also addresses specific case of #9019 but the more generic issue remains.

Asm diffs indicate also the following benefits: reduction in stack frame size, eliminating the need to zero init stack frame in some cases or reduce the size of zero init, totally eliminates some structs on frame due to copy prop.

The regressions are the cases where the promoted struct is getting passed as a parameter to a call.  A copy of promoted struct is made, which would be a field by field assignment instead of 16-byte read and 16-byte write if we had not promoted the struct.

Here are the SPMI diffs on CQPerf and Fx assemblies:

CqPerf suite:
```
09.70:         Code Size improvements (code sizes in bytes):
09.70:         <filename>                                            baseline        diff improvement  %improvement    #funcs
09.70:         CQ_PERF_Clean_Thin_Unique_all.dasm                        4765        4588         177          3.71        10
09.70:            6 functions improved (205 total bytes improvement)
09.70:               Improvement # 1 goes from   273 to   225, diff of    48 (17.58%) :: KNucleotide:KFrequency():this
09.70:               Improvement # 2 goes from   723 to   666, diff of    57 ( 7.88%) :: Microsoft.CodeAnalysis.CSharp.InContainerBinder:LookupSymbolsInSingleBinder(ref,ref,int,ref,int,ref,bool,byref):this
09.70:               Improvement # 3 goes from   160 to   134, diff of    26 (16.25%) :: Platform.IO.Data.SharedData:GetLongestSpanAt(long,int):struct:this
09.70:               Improvement # 4 goes from  1409 to  1352, diff of    57 ( 4.05%) :: Microsoft.CodeAnalysis.CSharp.Binder:LookupMembersInClass(ref,ref,ref,int,ref,int,ref,bool,byref):this
09.70:               Improvement # 5 goes from   418 to   407, diff of    11 ( 2.63%) :: Microsoft.CodeAnalysis.CSharp.LocalScopeBinder:LookupSymbolsInSingleBinder(ref,ref,int,ref,int,ref,bool,byref):this
09.70:               Improvement # 6 goes from   799 to   793, diff of     6 ( 0.75%) :: Microsoft.CodeAnalysis.CSharp.Binder:LookupAttributeType(ref,ref,ref,int,ref,int,bool,byref):this
09.70:            4 functions regressed (28 total bytes regression)
09.70:               Regression # 1 goes from   281 to   292, diff of    11 ( 3.77%) :: Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamedTypeSymbol:get_Layout():struct:this
09.70:               Regression # 2 goes from    59 to    64, diff of     5 ( 7.81%) :: Microsoft.CodeAnalysis.Common.CommonChildSyntaxList:GetEnumerator():struct:this
09.70:               Regression # 3 goes from   390 to   401, diff of    11 ( 2.74%) :: AssemblyDataForAssemblyBeingBuilt[__Canon,__Canon][System.__Canon,System.__Canon]:BindAssemblyReferences(struct,ref):ref:this
09.70:               Regression # 4 goes from   253 to   254, diff of     1 ( 0.39%) :: Microsoft.CodeAnalysis.CommonReferenceManager`2[__Canon,__Canon][System.__Canon,System.__Canon]:ResolveReferencedAssemblies(struct,struct,ref,bool):ref
09.70:         TOTAL                                                     4765        4588         177          3.71        10

```
Fx Assemblies:

```
02.82:         Code Size improvements (code sizes in bytes):
02.82:         <filename>                                            baseline        diff improvement  %improvement    #funcs
02.82:         All_FX_Clean_Thin_Unique_all.dasm                        47642       45938        1704          3.58       107
02.82:            48 functions improved (2330 total bytes improvement)
02.82:               Improvement # 1 goes from  1807 to  1324, diff of   483 (26.73%) :: System.Numerics.Matrix4x4:CreateConstrainedBillboard(struct,struct,struct,struct,struct):struct
02.82:               Improvement # 2 goes from   764 to   531, diff of   233 (30.50%) :: System.Numerics.Matrix4x4:CreateWorld(struct,struct,struct):struct
02.82:               Improvement # 3 goes from   867 to   647, diff of   220 (25.37%) :: System.Numerics.Matrix4x4:CreateBillboard(struct,struct,struct,struct):struct
02.82:               Improvement # 4 goes from   919 to   764, diff of   155 (16.87%) :: System.Numerics.Matrix4x4:CreateLookAt(struct,struct,struct):struct
02.82:               Improvement # 5 goes from   105 to    58, diff of    47 (44.76%) :: System.Numerics.Vector3:Negate(struct):struct
02.82:               Improvement # 6 goes from  1183 to  1057, diff of   126 (10.65%) :: System.ServiceModel.Activities.WorkflowService:CreateOperationProperties(ref):ref:this
02.82:               Improvement # 7 goes from  1822 to  1670, diff of   152 ( 8.34%) :: System.Numerics.Matrix4x4:Decompose(struct,byref,byref,byref):bool
02.82:               Improvement # 8 goes from   207 to   156, diff of    51 (24.64%) :: QName:Equals(ref):bool:this
02.82:               Improvement # 9 goes from    82 to    52, diff of    30 (36.59%) :: System.Numerics.Vector3:Multiply(struct,float):struct
02.82:               Improvement #10 goes from    83 to    55, diff of    28 (33.73%) :: System.Numerics.Vector3:Multiply(float,struct):struct
02.82:            1 aggregated functions improved (7 total bytes improvement)
02.82:               Improvement # 1 goes from   144 to   137, diff of     7 ( 4.86%) ::  2 x System.Xml.Linq.XContainer:AddFirst(ref):this
02.82:            35 functions regressed (633 total bytes regression)
02.82:               Regression # 1 goes from   188 to   229, diff of    41 (17.90%) :: System.Collections.Generic.ArraySortHelper`1[XamlNode][System.Xaml.XamlNode]:SwapIfGreater(ref,ref,int,int)
02.82:               Regression # 2 goes from   188 to   223, diff of    35 (15.70%) :: System.Collections.Generic.ArraySortHelper`1[ArraySegment`1][System.ArraySegment`1[System.Byte]]:SwapIfGreater(ref,ref,int,int)
02.82:               Regression # 3 goes from   188 to   223, diff of    35 (15.70%) :: System.Collections.Generic.ArraySortHelper`1[Section][System.IdentityModel.XmlBuffer+Section]:SwapIfGreater(ref,ref,int,int)
02.82:               Regression # 4 goes from   188 to   223, diff of    35 (15.70%) :: System.Collections.Generic.ArraySortHelper`1[Section][System.ServiceModel.XmlBuffer+Section]:SwapIfGreater(ref,ref,int,int)
02.82:               Regression # 5 goes from   164 to   188, diff of    24 (12.77%) :: System.Collections.Generic.ArraySortHelper`1[AlternativeIssuerEndpoint][System.ServiceModel.Security.Tokens.IssuedSecurityTokenParameters+AlternativeIssuerEndpoint]:Swap(ref,int,int)
02.82:               Regression # 6 goes from   129 to   150, diff of    21 (14.00%) :: System.Collections.Generic.ArraySortHelper`1[XamlNode][System.Xaml.XamlNode]:Swap(ref,int,int)
02.82:               Regression # 7 goes from   573 to   614, diff of    41 ( 6.68%) :: System.ServiceModel.Channels.DefaultWebSocketConnectionHandler:ParseSubprotocolValues(ref):struct
02.82:               Regression # 8 goes from   251 to   276, diff of    25 ( 9.06%) :: System.Collections.Generic.ArraySortHelper`1[AlternativeIssuerEndpoint][System.ServiceModel.Security.Tokens.IssuedSecurityTokenParameters+AlternativeIssuerEndpoint]:SwapIfGreater(ref,ref,int,int)
02.82:               Regression # 9 goes from   129 to   145, diff of    16 (11.03%) :: System.Collections.Generic.ArraySortHelper`1[ArraySegment`1][System.ArraySegment`1[System.Byte]]:Swap(ref,int,int)
02.82:               Regression #10 goes from   129 to   145, diff of    16 (11.03%) :: System.Collections.Generic.ArraySortHelper`1[Sectio
```

### CODE REVIEW

**Comment by sivarv** [comment]
@briansull  - Please review this 
CC @dotnet/jit-contrib

**Comment by CarolEidt** [comment]
> The regressions are the cases where the promoted struct is getting passed as a parameter to a call. A copy of promoted struct is made, which would be a field by field assignment instead of 16-byte read and 16-byte write if we had not promoted the struct.

It would be good to consider a heuristic that considers the impact of passing a struct local as a call and of assigning it the value returned from a call. This, of course, will have to be ABI-specific (because for some combinations of ABI and struct it would be more advantageous, rather than less, for it to have been promoted). However, this would require the JIT to keep track of this information, presumably during the importer since struct promotion happens fairly early.

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -17132,9 +17132,12 @@ void Compiler::fgPromoteStructs()
                 // field accesses, but only block-level operations on the whole struct, if the struct
                 // has only one or two fields, then doing those block operations field-wise is probably faster
                 // than doing a whole-variable block operation (e.g., a hardware "copy loop" on x86).
-                // So if no fields are accessed independently, and there are three or more fields,
+                // Struct promotion also provides the following benefits: reduce stack frame size,
+                // reduce the need for zero init of stack frame, constant prop and copy prop.
+                // Asm diffs indicate that promoting structs up to 3 fields is a net size win.
```
Comment (CarolEidt) [inline src/jit/morph.cpp]: You might reword "constant prop and copy prop" to be "finer-grained constant prop and copy prop" as we support both for full structs, but only for the entire value.
You might also note here (as well as possibly opening another issue, if such doesn't already exist) that ideally we would want to consider the impact of passing and returning struct values to/from calls.

**Comment by briansull** [comment]
LGTM

File: src/jit/morph.cpp (hunk 1 above)
Comment (sivarv) [inline src/jit/morph.cpp]: I have updated comments and also opened an issue (https://github.com/dotnet/coreclr/issues/9477) to track it

**Pull request merged**